src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,206 @@
     public void add(T value) {
+        if (isEmpty()) {
+            linkEmptyList(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(tail, value);
+        newNode.prev.next = newNode;
+        tail = newNode;
+        newNode.prev = tail.prev;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0 && size == 0) {
+            linkEmptyList(value);
+            return;
+        }
+        if (index == 0) {
+            linkHead(value);
+            return;
+        }
+        if (index == size) {
+            linkTail(value);
+            return;
+        }
+        checkIndexOutOfBounds(index);
+        Node<T> current = getNode(index);
+        Node<T> newNode = new Node<>(current.prev,value,current);
+        current.prev.next = newNode;
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.size() == 0) {
+            return;
+        }
+        for (T element:list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexOutOfBounds(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexOutOfBounds(index);
+        final Node<T> current = getNode(index);
+        final T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBounds(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (object.equals(x.value)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkEmptyList(T value) {
+        Node<T> newNode = new Node<>(value);
+        head = newNode;
+        tail = newNode;
+        size++;
+    }
+
+    private void linkHead(T value) {
+        final Node<T> f = head;
+        final Node<T> newNode = new Node<>(value, f);
+        head = newNode;
+        f.prev = newNode;
+        size++;
+    }
+
+    private void linkTail(T value) {
+        Node<T> l = tail;
+        Node<T> newNode = new Node<>(l, value);
+        tail = newNode;
+        l.next = newNode;
+        size++;
+    }
```

Comment:
```suggestion
        Node<T> newNode = new Node<>(value, head);
        head.prev = newNode;
        head = newNode;
        size++;
    }

    private void linkTail(T value) {
        Node<T> newNode = new Node<>(tail, value);
        tail.next = newNode;
        tail = newNode;
        size++;
```
without common "size++" line logic only takes 3 lines, and may be not worth extracting to a private method, but it's ok and up to you